pkg/cmd/infra/deployer/deployer.go

```diff
@@ -72,5 +87,75 @@ func deploy(cfg *config) error {
 		ReplicationController: strategy.RealReplicationController{KubeClient: kClient},
 		Codec: latest.Codec,
 	}
-	return strategy.Deploy(deployment)
+
+	return strategy.Deploy(newDeployment, oldDeployments)
+}
+
+// getDeployerContext finds the target deployment and any deployments it considers to be prior to the
+// target deployment. Only deployments whose LatestVersion is less than the target deployment are
+// considered to be prior.
+func getDeployerContext(controllerGetter replicationControllerGetter, namespace, deploymentName string) (*kapi.ReplicationController, []kapi.ObjectReference, error) {
+	var err error
+	var newDeployment *kapi.ReplicationController
+	var newConfig *deployapi.DeploymentConfig
+
+	// Look up the new deployment and its associated config.
+	if newDeployment, err = controllerGetter.Get(namespace, deploymentName); err != nil {
+		return nil, nil, err
+	}
+
+	if newConfig, err = deployutil.DecodeDeploymentConfig(newDeployment, latest.Codec); err != nil {
+		return nil, nil, err
+	}
+
+	glog.Infof("Found new deployment %s for config %s with latestVersion %d", newDeployment.Name, newConfig.Name, newConfig.LatestVersion)
+
+	// Collect all deployments that predate the new one by comparing all old ReplicationControllers with
+	// encoded DeploymentConfigs to the new one by LatestVersion. Treat a failure to interpret a given
+	// old deployment as a fatal error to prevent overlapping deployments.
+	var allControllers *kapi.ReplicationControllerList
+	oldDeployments := []kapi.ObjectReference{}
+
+	if allControllers, err = controllerGetter.List(newDeployment.Namespace, labels.Everything()); err != nil {
+		return nil, nil, fmt.Errorf("Unable to get list replication controllers in deployment namespace %s: %v", newDeployment.Namespace, err)
+	}
+
+	glog.Infof("Inspecting %d potential prior deployments", len(allControllers.Items))
+	for _, controller := range allControllers.Items {
+		if configName, hasConfigName := controller.Annotations[deployapi.DeploymentConfigAnnotation]; !hasConfigName {
+			glog.Infof("Disregarding controller %s which doesn't appear to be a deployment", controller.Name)
+			continue
+		} else if configName != newConfig.Name {
+			glog.Infof("Disregarding deployment %s for unrelated config %s", controller.Name, configName)
+			continue
+		}
+
+		if oldVersion, convErr := strconv.Atoi(controller.Annotations[deployapi.DeploymentVersionAnnotation]); convErr != nil {
+			return nil, nil, fmt.Errorf("Couldn't determine version of deployment %s: %v", controller.Name, convErr)
```

Comment:
Unless you're relying on the `if` block scope in the else. Otherwise I'd have to pre-declare `oldVersion`. Which do you prefer?